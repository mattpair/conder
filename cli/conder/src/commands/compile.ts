
import { GluegunToolbox, GluegunCommand } from 'gluegun'
import {string_to_environment} from 'conduit_compiler'  

const command: GluegunCommand = {
  name: 'compile',
  alias: ['c'],
  run: async (toolbox: GluegunToolbox) => {
    const {
      parameters,
      print: { info, error },
      filesystem,
    } = toolbox
    
    const deployment_name = parameters.first === undefined ? "app" : parameters.first
    const artifact_name = `${deployment_name}.json`
    info(`Writing artifacts to ${artifact_name}`)
    
    const conduit = filesystem.read("main.cdt")
    if (conduit === undefined) {
      error(`Could not find required file main.cdt`)
      process.exit(1)
    }
    const output = string_to_environment(conduit)
    switch (output.kind) {
      case "success":
        filesystem.write(artifact_name, JSON.stringify(output.env))
        break
      case "error":
        error(output.reason)
        process.exit(1)
    }
  }
}

module.exports = command
