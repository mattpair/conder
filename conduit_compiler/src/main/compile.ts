
import { Parse} from "./parse";
import { FunctionResolved, TypeResolved, Message, Field } from "./entity/resolved";
import { resolveDeps } from "./resolution/resolveTypes";
import { FileLocation } from "./util/filesystem";
import { Enum, EnumMember } from "./entity/basic";
import { resolveFunctions } from "./resolution/resolveFunction";

export type PartiallyCompiled = Record<string, {proto: string, data: FunctionResolved.File}>
export function compileFiles(files: Record<string, () => string>): PartiallyCompiled {
    const conduits: Parse.File[] = []
    for (const file in files) {
        if (file.endsWith(".cdt")) {
            conduits.push(Parse.extractAllFileEntities(files[file](), new FileLocation(file)))
        }
    }

    const r = resolveDeps(conduits)

    return resolveFunctionsAndProto(r)
} 

function resolveFunctionsAndProto(files: TypeResolved.File[]): PartiallyCompiled {
    const results: PartiallyCompiled = {}
    files.forEach(file => {

        const enums: Enum[] = []
        const messages: Message[] = []
        file.inFileScope.forEach(v => {
            switch(v.kind) {
                case "Enum":
                    enums.push(v)
                    break;
                case "Message":
                    messages.push(v)
                    break;
            }
        })
        let proto = ''
        if (enums.length + messages.length > 0) {
            proto = `
syntax="proto2";
${file.children.Import.map(d => `import "${d.dep.replace(".cdt", ".proto")}";`).join("\n")}
            
${enums.map(printEnum).join("\n\n")}
${messages.map(printMessage).join("\n\n")}
`
        }

        results[`${file.loc.fullname.replace(".cdt", ".proto")}`] = {proto, data: resolveFunctions(file)}
        
    })

    return results       
}

function printEnum(e: Enum): string {
    const mems = printMembers(e.children.EnumMember)
    return `
enum ${e.name} {
${mems}
}
    `
}

function printMembers(m: EnumMember[]): string {
    return m.map((e, index) => `\t${e.name} = ${index + 1};`).join("\n")
}

function printMessage(m: Message): string {
    const fields = printFields(m.children.Field)

    return `
message ${m.name} {
${fields}
}`
}

function printFields(fields: Field[]): string {
    return fields
    .map((f, index) => {
        const p = f.part.FieldType.differentiate()
        return `\t${f.isRequired ? 'required' : 'optional'} ${p.kind === "Primitive" ? p.val : p.name} ${f.name} = ${index + 1};`
    })
    .join("\n")
}