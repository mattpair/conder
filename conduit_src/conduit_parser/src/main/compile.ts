
import { Parse} from "./parse";
import { Manifest, Python3Install, EntityMap } from "./entity/resolved";
import { ConduitBuildConfig } from "./entity/ConduitBuildConfig";
import { toEntityMap, PartialEntityMap } from "./resolution/typeValidation";
import { FileLocation } from "./utils";
import { generateSystemObjects,  } from "./resolution/generateSystemStructs";

export function compileFiles(files: Record<string, () => string>, build: ConduitBuildConfig): Manifest {
    const conduits: Parse.File[] = []
    for (const file in files) {
        conduits.push(Parse.extractAllFileEntities(files[file](), new FileLocation(file)))
    }
    const map: PartialEntityMap<Python3Install>  = generateSystemObjects(toEntityMap(conduits))
    
    if (build.install) {
        build.install.forEach(i => {
            const g = map.get(i.name)
            if(g) {
                throw Error(`Install name ${i.name} collides with ${g.kind} ${g.name}`)
            }
            map.set(i.name, i)
        })
    }
    
    return {inScope: new EntityMap(map)}
} 