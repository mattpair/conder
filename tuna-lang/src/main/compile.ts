import { ParseResult, globals_$0_$0, ASTKinds } from "./parser";


export function compile(p: ParseResult): globals_$0_$0[] {
    if (p.err) {
        throw Error(`Failure parsing: line ${p.err.pos.line} col ${p.err.pos.offset}`)
    }
    return p.ast.map(g => {
        switch (g.value.kind) {
            case ASTKinds.varDecl: 
                if (g.value.mutability !== "const") {
                    throw Error(`Global variable ${g.value.name.name} must be const`)
                }
            case ASTKinds.func: 

                return g.value

            default: 
                const ne: never = g.value
        }
    })

}