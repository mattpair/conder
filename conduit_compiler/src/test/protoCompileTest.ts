import compile from "../main/compileToProto"


function protoCompileTest(description, file: string) {
    test(description, () => {
        expect(compile(file)).toMatchSnapshot()
    })
} 

protoCompileTest("should convert simple conduit into proto2",
`
    message MyMessage {
        double d,
        float f,
        int32 i32,
        int64 I,
        uint32 u32,
        uint64 u64,
        sint32 s32,
        sint64 s64,
        fixed32 f32,
        fixed64 f64,
        sfixed32 sf32,
        sfixed64 sf64,
        bool b,
        string s,
        bytes bs,
    }`
)

protoCompileTest("should convert multiple conduit into proto2", `
    message m1 {
        double d,
        float f,
        int32 i32,
        int64 I,
        uint32 u32,
        uint64 u64,
    }
    message m2{
        sint32 s32,
        sint64 s64,
        fixed32 f32,
        fixed64 f64,
        sfixed32 sf32,
        sfixed64 sf64,
        bool b,
        string s,
        bytes bs,
    }
    `)

protoCompileTest("should allow one message to reference another", `
    message m1 {
        double d,
        float f,
        int32 i32,
        int64 I,
        uint32 u32,
        uint64 u64,
    }
    message m2{
        sint32 s32,
        m1 myMessag,
    }
    `)

protoCompileTest("should allow the specification of required fields", `
    message MyMessage {
        required double d,
        required float f,
        required int32 i32,
        required int64 I,
        required uint32 u32,
        required uint64 u64,
        required sint32 s32,
        required sint64 s64,
        required fixed32 f32,
        required fixed64 f64,
        required sfixed32 sf32,
        required sfixed64 sf64,
        required bool b,
        required string s,
        required bytes bs,
    }
    `)

protoCompileTest("should allow enums",`
    enum PurchaseCategory {
        UNCATEGORIZED,
        GAS,
        DINING,
        TRAVEL,
        GROCERY,
        ENTERTAINMENT,
    }
    `)