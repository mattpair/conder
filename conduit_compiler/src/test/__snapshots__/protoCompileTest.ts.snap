// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should allow enums 1`] = `
"
    
enum PurchaseCategory {
	UNCATEGORIZED = 1;
	GAS = 2;
	DINING = 3;
	TRAVEL = 4;
	GROCERY = 5;
	ENTERTAINMENT = 6;
}
    
    
    "
`;

exports[`should allow enums separated by whitespace 1`] = `
"
    
enum PurchaseCategory {
	UNCATEGORIZED = 1;
	GAS = 2;
	DINING = 3;
	TRAVEL = 4;
	GROCERY = 5;
	ENTERTAINMENT = 6;
}
    
    
    "
`;

exports[`should allow new line to end field 1`] = `
"
    
    
message MyMessage {
	required double d = 1;
	optional float f = 2;
	required int32 i32 = 3;
	required int64 I = 4;
	required uint32 u32 = 5;
	required uint64 u64 = 6;
	optional sint32 s32 = 7;
	required sint64 s64 = 8;
	required fixed32 f32 = 9;
	required fixed64 f64 = 10;
	required sfixed32 sf32 = 11;
	required sfixed64 sf64 = 12;
	required bool b = 13;
	required string s = 14;
	required bytes bs = 15;
}
    "
`;

exports[`should allow one message to reference another 1`] = `
"
    
    
message m1 {
	optional double d = 1;
	optional float f = 2;
	optional int32 i32 = 3;
	optional int64 I = 4;
	optional uint32 u32 = 5;
	optional uint64 u64 = 6;
}


message m2 {
	optional sint32 s32 = 1;
	optional m1 myMessag = 2;
}
    "
`;

exports[`should allow the specification of required fields 1`] = `
"
    
    
message MyMessage {
	required double d = 1;
	required float f = 2;
	required int32 i32 = 3;
	required int64 I = 4;
	required uint32 u32 = 5;
	required uint64 u64 = 6;
	required sint32 s32 = 7;
	required sint64 s64 = 8;
	required fixed32 f32 = 9;
	required fixed64 f64 = 10;
	required sfixed32 sf32 = 11;
	required sfixed64 sf64 = 12;
	required bool b = 13;
	required string s = 14;
	required bytes bs = 15;
}
    "
`;

exports[`should convert multiple conduit into proto2 1`] = `
"
    
    
message m1 {
	optional double d = 1;
	optional float f = 2;
	optional int32 i32 = 3;
	optional int64 I = 4;
	optional uint32 u32 = 5;
	optional uint64 u64 = 6;
}


message m2 {
	optional sint32 s32 = 1;
	optional sint64 s64 = 2;
	optional fixed32 f32 = 3;
	optional fixed64 f64 = 4;
	optional sfixed32 sf32 = 5;
	optional sfixed64 sf64 = 6;
	optional bool b = 7;
	optional string s = 8;
	optional bytes bs = 9;
}
    "
`;

exports[`should convert simple conduit into proto2 1`] = `
"
    
    
message MyMessage {
	optional double d = 1;
	optional float f = 2;
	optional int32 i32 = 3;
	optional int64 I = 4;
	optional uint32 u32 = 5;
	optional uint64 u64 = 6;
	optional sint32 s32 = 7;
	optional sint64 s64 = 8;
	optional fixed32 f32 = 9;
	optional fixed64 f64 = 10;
	optional sfixed32 sf32 = 11;
	optional sfixed64 sf64 = 12;
	optional bool b = 13;
	optional string s = 14;
	optional bytes bs = 15;
}
    "
`;
